etl/scripts-ccao-data-warehouse-us-east-1/spatial/spatial-parcel.R

```diff
@@ -122,40 +148,67 @@ process_parcel_file <- function(s3_bucket_uri,
       # For each PIN10, keep the centroid of the largest polygon
       group_by(pin10) %>%
       arrange(desc(area)) %>%
-      summarize(across(c(lon, lat, x_3435, y_3435), first)) %>%
+      summarize(across(c(lon, lat, x_3435, y_3435), dplyr::first)) %>%
       ungroup()
     tictoc::toc()
 
     # Read attribute data and get unique attributes by PIN10
-    tictoc::tic(paste("Joined and wrote parquet for:", file_year))
+    tictoc::tic(paste("Merged boundary and attribute data:", file_year))
     attr_df <- read_parquet(local_attr_file) %>%
       mutate(pin10 = str_sub(pin, 1, 10)) %>%
       group_by(pin10) %>%
       summarize(
-        across(c(tax_code, nbhd_code, town_code), first),
+        across(c(tax_code, nbhd_code, town_code), dplyr::first),
         year = file_year
       ) %>%
       ungroup()
 
-    # Merge spatial boundaries with attribute data
+    # Merge spatial boundaries with attribute data and drop sliver polygons
     spatial_df_merged <- spatial_df_clean %>%
       left_join(attr_df, by = "pin10") %>%
       left_join(spatial_df_centroids, by = "pin10") %>%
       mutate(
         has_attributes = !is.na(town_code),
         geometry = st_transform(geometry, 4326),
-        geometry_3435 = st_transform(geometry, 3435)
+        area = units::drop_units(st_area(geometry))
       ) %>%
+      # Drop anything with an area less than 1 sq meter, as long as it has other
+      # polygons for the same PIN10 that are larger
+      filter(!(area <= 1 & n() >= 2 & !all(area <= 1)), .by = "pin10") %>%
       select(
         pin10, tax_code, nbhd_code, has_attributes,
-        lon, lat, x_3435, y_3435, geometry, geometry_3435,
+        lon, lat, x_3435, y_3435, geometry,
         town_code, year
-      ) %>%
-      distinct(pin10, .keep_all = TRUE)
+      )
+    tictoc::toc()
+
+    # Aggregate PIN10s with more than one row using st_union. Split them out
+    # from the rest of the dataset because st_union is an incredibly expensive
+    # operation, so we only want to perform it on the PINs we need to
+    tictoc::tic(paste("Collapsed parcels to the PIN10 level:", file_year))
+    spatial_df_merged <- bind_rows(
+      spatial_df_merged %>%
+        filter(n() >= 2, .by = pin10) %>%
+        summarize(
+          across(tax_code:year, dplyr::first),
+          across(geometry, st_union),
+          .by = "pin10"
+        ),
+      spatial_df_merged %>%
+        filter(!(n() >= 2), .by = pin10)
+    ) %>%
+      st_cast("MULTIPOLYGON") %>%
+      mutate(geometry_3435 = st_transform(geometry, 3435)) %>%
+      relocate(geometry_3435, .after = geometry)
+    tictoc::toc()
```

Comment:
Here we aggregate to the PIN10 level using a geometry union instead of `distinct()`. This also improves the feature results quite a bit, as they now include all polygons.